### DISCUSSION TITLE
Sorting and scan a large number of arrays of varying sizes

### DISCUSSION THREAD

**Discussion by wanghongyu2001** [NVIDIA/cccl#1587] [CUB]
I want to use the CUB library for sorting, with different blocks handling arrays of different sizes. However, I need to specify the parameters of ```cub::BlockRadixSort``` at compile time. But I can only determine the sizes of these arrays at runtime, as each block handles arrays of different sizes. So, it seems that I cannot use CUB in this case. Is there any solution? Or should I use batch Thrust sort instead?

  For example, if I need to sort and compute prefix sums for hundreds of thousands of arrays, with sizes ranging from 0 to several hundred thousand, is there any way to efficiently perform these operations on the GPU using the CUB library?

Any suggestions are helpful.

**Comment by elstehle** [comment]
Thank you for reaching out with your question, @wanghongyu2001.

> For example, if I need to sort and compute prefix sums for hundreds of thousands of arrays, with sizes ranging from 0 to several hundred thousand, is there any way to efficiently perform these operations on the GPU using the CUB library?

If your arrays reside at different offsets within the same memory allocation, you can use [cub::DeviceSegmentedSort](https://nvidia.github.io/cccl/cub/api/structcub_1_1DeviceSegmentedSort.html?highlight=devicesegmentedsort).

Here's an [godbolt example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIAruiakl9ZATwDKjdAGFUtEywYgAzAA5SjgBk8BkwAOXcAI0xiCQAmUgAHVAVCOwYXNw9vPySU2wEgkPCWKJiueKtMGzShAiZiAgz3T19LTGt8hlr6gkKwyOi4yzqGpqzWhRHe4P6SwfKASktUE2Jkdg4AUlivYOQ3LABqTa8nUwiAenOAOlMEE%2BxNjQBBJ9edvYPMY9OCBGITJMLlgAG54dYAfRBVSIxGu9y8jxe212DH2Zm%2BJycfwBQIQyQIUJhJHhDzeb2CBEOLCYwQglMO9WAyFIh2QCHqhwAVFymSCFuSAOwAITeh3Fh0BwWAh3MdQhVJOABFDpSTqKXhLDhcLochKwEvRZUw6mKJQzDgx3BDCJgWAofirBeqzeKLVaWBClMA2IIHcrDrEXZqJZN0CAQNCbCQsWrEYdUFQqEoCP6vCrNiKNKyvKznYKlcHnlqwxGo7DY4IHocANaYACeChtDEdxxFfkOADY86yAKw51nZw4ATkzha8GuLEtd2t1ymImAS9W%2BoPBmAAtCkjnKmDOcYCCBHV5DyzHTjuFdX0BDE8nMKnM8LbymFGOi1r95Mj5gwSficQsQvAgrwhOtG2bR8wKbYI3wnPd/gPb9f0wIlowA88TSYS942vKCbxMYCRTw4JrhSAAvTAIAFAsixnIDuVwhtoIYCEEgIYhW0/Q8QGIJgAHdWNQSloghURJggRjwJIncqIFOCQ3FeiuUkpsVkJNiOIDLiI14gSkmE4hRKYcSVPwghrhkhY5MnGcdUOJV72iFgZkOAg7SSXjiHrWUfzXQ5JhIJhgG%2BBcAEcTDwBdfQfBTDhBIT0C1BiFXcr1YSCzF00OUI5ACAJ31DPAKIVCU3JYBI0sC4KIQies3LTFUNAK8VzgjBzkKETAfUYNz0CEEhuP6hoAGkmIgGctWvMqKoC3jgtZabKrmlDavq0gJolUzglY9jWVMtSdrIY5Yq1D0bTKhRWTO71osu2UbyTF9biiYA6SWe7n3vBRnq6t7jliYVDi4azyViuznloWhUFENzXPcwKvP89LgpnUxzAAWTEKHkHG2JOym1LZoyhbCeRla6swBQQeRMHdQAJRMFsFAG6UEwSaITTSVGTAiNrfPWTrusETA%2BoGiMhoIUbG3Gk7NpS8qluJuGFaJ6rVsp9bZcU0CmObQ69p18CDo01kNvFM7bXtK7rRunq7uvT7Ux%2B16GCog3He%2B5AXr%2B7ZAeB5rjqncU7IAcXvQ4FwUExaFTf7O0OBJiDVWKmAI1BLUwVRCVtwRztbD3nbegPtJAfFJlQisMPlYD4wQL0BpFszH32gjYMnLV%2BGICBSxAciUKpPBlSaidVR%2BJxDjr5mGkbtTSKKyjrNVX3feo9ukrdKh6UdAMuRCLOvS66LztXs2kp7tACLHrF/qDXsnBbddH71Q%2BesOb2GAYVmTQTR7w%2B2WIr6nG5HvbOL9c6EEAePf%2BhxH7rmOHfBg/8A7r3NpnUBQtCSEGXgDZB04aJayRuGFAalIGjyxNff%2BIAb6kMng3B2rdezCkHr2DMpwKGxFvvfJB8kg5tkLLFBcBBVgtmHjZAsHAli0E4L2XgngOBaFIKgTgTg5BKmLE4bBgNmarHWP9LwPBSCHnkRIpYNYQC9i4NcWIgpJBeFiJISQnZJDDkkD4Zx%2BhOCSFkZoXgSiOC8AUCAbMRitBLDgLAGAiBiHlToNEcglA0AxPoDEBgIJkDIC4FwPwNAY7RECRAXmxjSARGCPUesnADElOYF5AA8hEbQMIKm8ESdFGpDBaDlKKVgCIJhgBOCxoE7gvAsA0iMOILpkUYR4GhIMhRmcqgEQ2Ao4SUiim0DwBETy9YXBYB8YYpOLAmmkGhMQCIyRMAOVGcAdZRgfFLCoAYYACgABqeBMB8RqezORBj%2BCCBEGIdgUgZCCEUCodQRTdBcH0IYYwZgLDrIiIEyASxUBsTSIM9cwU0QYG%2BPUdkypEkJEWRCQU2Y0BYGVAoT0pKYE1K8L4k5ScsBIrdpUao9gICODGJ4KFgQZjFFKHoXIqQBDcqFckEVDA%2BgCvmG0DoNQphiqhWyzo3QGjSoGGUYYPQlXavVfyzVEgljaLWICyR0jvFFL8dlZ5Thx6ZOuD4a4GhDgQFwIQEgejga8BCSY0gZivBeGuIGkNoaw0eI4F40ghzezZjkQovxASgmGLuaQcJUS1JEoIPEiAhLYnEFCKwDYoRbX2tiM664BiRYeqZXoX5whRDiCBfW0Fag9mQtIHxXiCQmnmo4DI0g8bfGcBqQRLNP8bV2qBpYp1LqIAuCSdEL1CwfV3KWAgTATAsAxDdqsqNMa417MTZYZNvqrL%2BpAJISxXhBSdl7C4rgpKOGxC4Bobsqz6WDqPZwVdxiwmRIgEgY8mAc15uSYWtgnAS1TodbO4Z/MNi8GrbCPA4YoX1v%2BU26QLalBtohXoFVaQHAMGcK4Zoeg%2BVFENVC4VnRdW0bSBquYWrCMCDVY0MjWRlXtCmWxqYTHBXKsVZxnlerphUeY0a/ZmBMCoaCX2gdQ7FGcHan5V57yl3QftTO51rr3Wwi9ayBdCR83Lt/aE0xIBBTBucRoHwXAvBOJ8LEYcXhhyCgjVGpTx7AnBLXRG2IlqE0/pTX%2BpYJyUj2EkEAA%3D) that you can play around with.

<details>
<summary>Usage example</summary>

```
#include <cub/cub.cuh>

#include <thrust/device_vector.h>
#include <thrust/host_vector.h>

int main(int argc, char **argv)
{
    using data_t = int;
    // Sample data
    int  num_items = 7;
    int  num_segments = 2;
    std::vector<int> offsets = {0, 3, 7};
    std::vector<int> keys_in = {8, 6, 7, 5, 3, 0, 9};
    
    // Prepare device-side data
    thrust::device_vector<data_t> d_offsets{offsets};
    thrust::device_vector<data_t> d_keys_in{keys_in};
    thrust::device_vector<data_t> d_keys_out{keys_in.size()};

    data_t *d_keys_in_ptr = thrust::raw_pointer_cast(d_keys_in.data());
    data_t *d_keys_out_ptr = thrust::raw_pointer_cast(d_keys_out.data());

    // Determine temporary device storage requirements
    void     *d_temp_storage = NULL;
    size_t   temp_storage_bytes = 0;
    cub::DeviceSegmentedSort::SortKeys(
        d_temp_storage, temp_storage_bytes,
        d_keys_in_ptr, d_keys_out_ptr, 
        num_items, num_segments, d_offsets.cbegin(), d_offsets.cbegin() + 1);

    // Allocate temporary storage
    cudaMalloc(&d_temp_storage, temp_storage_bytes);

    // Run sorting operation
    cub::DeviceSegmentedSort::SortKeys(
        d_temp_storage, temp_storage_bytes,
        d_keys_in_ptr, d_keys_out_ptr, 
        num_items, num_segments, d_offsets.cbegin(), d_offsets.cbegin() + 1);
    
    // Get results & print
    auto next_segment_it = offsets.cbegin();
    thrust::host_vector<data_t> h_sorted_out{d_keys_out};
    for(std::size_t i=0; i < h_sorted_out.size(); i++){
        if(i == *next_segment_it){
            std::cout << "\n --- Segment beginning at offset " << *next_segment_it << " --- \n";
            next_segment_it++;
        }
        std::cout << i << ": " << h_sorted_out[i] << "\n";
    }
    return 0;
}
```

Output:
```
 --- Segment beginning at offset 0 --- 
0: 6
1: 7
2: 8

 --- Segment beginning at offset 3 --- 
3: 0
4: 3
5: 5
6: 9
```

</details>

Since you also asked about prefix sum, could you elaborate what your to-be-prefix-summed input looks like?

**Reply by wanghongyu2001** [reply]
I want to implement sorting and deduplication for an array. 


In the CPP code, I use ```std::sort```, and ```std::map``` to implement these operations.
``` cpp
std::sort(tmp.begin(), tmp.end());

std::map<int, int> clean_map;
clean_map[tmp[0]] = 0;
int len = 1;
for (int j = 1; j < arr_size; j++)
{
    while (tmp[j] == tmp[j - 1] && j < arr_size - 1) j++;
    clean_map[tmp[j]] = len++;
}
```
In CUDA, I intend to use a bitmap for deduplication and utilize scan to obtain the indices.

Using a bitmap for deduplication and obtaining the index after sorting makes it relatively easy to implement in CUDA. Additionally, I need to obtain the indices after sorting and deduplication.

However, it requires a significant amount of memory, for example
```
key: 0 10000 20000
bit map size at least 20000.
```
In CUDA kernel
```cpp
// value range 0 - several hundred thousand
d_in: 0 4 2 1 4 2 2 1
//sort
d_out: 0 1 1 2 2 2 4 4
// map
for (int i = threadIdx.x; i < len; i += blockDim.x)
{
    map[d_out[i]] = 1;
}

//get map If map[i] == 1, it means this position has a number
map: 1 1 1 0 1

//scan Obtain a simple map
map: 0 1 2 3 3 4

// idx from 0 
map: 0 1 2 2 3
/*
key: 0 1 2 4
idx : 0 1 2 3
*/
//
for (int i = threadIdx.x; i < len; i += blockDim.x)
{
    int col = d_in[i];
    idx[i] = map[col];
}
```
Additionally, note that there are many arrays of different sizes that need to undergo this process.

Thank you once again for your help. This problem has been bothering me for many days.

**Reply by elstehle** [reply]
Thank you for providing an elaborate example. To be able to provide a more helpful answer I would need to understand how your arrays are arranged. 

Are the arrays within the same, contiguous memory allocation, just beginning at different offsets: 
```
// e.g.: array_0 begins at offset 0: 0, 4, 2
// e.g.: array_1 begins at offset 3: 1, 4, 2, 2, 1
int *my_arrays; // {0, 4, 2, 1, 4, 2, 2, 1};
```
or are they at distinct memory locations/buffers:
```
int *my_array[0] // {0, 4, 2};
int *my_array[1] // {3: 1, 4, 2, 2, 1};
```

**Reply by wanghongyu2001** [reply]
yes, the arrays within the same, contiguous memory allocation, just beginning at different offsets.

Then sort and remove duplicates for different segments, aiming to obtain the corresponding indices. Let's take an example of a segment.
```
keys:
0 1 2 100 0 1 2 101
```
I want to get the output:
```
out:
0 1 2 3 0 1 2 4
```
My approach is to first sort and remove duplicates to get:
```
key: 0 1 2 100 101
idx: 0 1 2 3   4
```
Then map back to the original array.

Thank you very much for your assistance.

**Reply by elstehle** [reply]
I think what you are trying to implement is similar to [cub::DeviceSelect::UniqueByKey](https://nvidia.github.io/cccl/cub/api/structcub_1_1DeviceSelect.html?highlight=uniquebykey#_CPPv4I000000EN3cub12DeviceSelect11UniqueByKeyE11cudaError_tPvR6size_t17KeyInputIteratorT19ValueInputIteratorT18KeyOutputIteratorT20ValueOutputIteratorT20NumSelectedIteratorT9NumItemsT12cudaStream_tb) or a [cub::DeviceRunLengthEncode](https://nvidia.github.io/cccl/cub/api/structcub_1_1DeviceRunLengthEncode.html?highlight=devicerunlengthencode). You may also want to take a look at https://github.com/NVIDIA/cucollections

Am I correct in assuming your input is given by a contiguous array and an offsets array? 
E.g., 
```
int *my_arrays; // {0, 4, 0, 1, 4, 2, 2, 1};
int *my_offsets; // {0, 3, 8}
```

And your desired output would return the index of the first occurrence of an item within its array? E.g., for two arrays of sizes 3 and 5:
```
0  1  2 | 3  4  5  6  7 <= indexes
0, 4, 0 | 1, 4, 2, 2, 1 <= input
0, 1, 0 | 3, 4, 5, 5, 7 <= output
```

**Reply by wanghongyu2001** [reply]
Thank you.

It's not like that; each segment is independent and does not affect others. The returned value is just the index of this segment, so it has high parallelism.
E. g.
```
0, 3, 8                 <= offset
0  1  2 | 3  4  5  6  7 <= indexes
0, 4, 0 | 1, 4, 2, 2, 1 <= input
0, 1, 0 | 0, 2, 1, 1, 0 <= output
```
If using cub::DeviceSelect::UniqueByKey, does it not support segments?

**Reply by elstehle** [reply]
I am not sure whether the output of the second segment in your example would is correct. Wouldn't it be:
```
0, 3, 8                 <= offset
0  1  2 | 3  4  5  6  7 <= indexes
0, 4, 0 | 1, 4, 2, 2, 1 <= input
0, 1, 0 | 0, 1, 2, 2, 0 <= output
```

> If using cub::DeviceSelect::UniqueByKey, does it not support segments?

No, unfortunately, `cub::DeviceSelect::UniqueByKey` does not support a segmented version of the algorithm yet.

**Reply by wanghongyu2001** [reply]
Yes, you are correct. I made a mistake.

```
0, 3, 8                 <= offset
0  1  2 | 3  4  5  6  7 <= indexes
0, 4, 0 | 1, 4, 2, 2, 1 <= input
0, 1, 0 | 0, 2, 1, 1, 0 <= output
```

So, is there an appropriate solution for this?

Thank you very much for your assistance.

**Reply by elstehle** [reply]
Aha, if I understand correctly, what you are trying to implement is a segmented order-preserving dictionary encoding?

There is an option to use `cub::DeviceSelect::UniqueByKey` but it will require a few extra steps, `thrust::zip_iterator`s during `cub::DeviceUniqueByKey`, etc. For simplicity, I am generating the absolute indexes into a "global" dictionary (similar to my original output suggestion example). It will only require some extra work to replace it with relative dictionary offsets. Here's a quick rundown of the data transformation steps:
```
0, 3, 8                 <= offset
0  1  2 | 3  4  5  6  7 <= indexes
0, 4, 0 | 1, 4, 2, 2, 1 <= input
---- Sort
0, 1, 2 | 3, 4, 5, 6, 7 <= sort_indexes (indexes of sorted items)
0, 0, 4 | 1, 1, 2, 2, 4 <= post_sort_items (items after segmented sort)
0, 2, 1 | 3, 7, 5, 6, 4 <= original_item_indexes (indexes after segmented sort)
---- Generate auxiliary data from offsets
0, 0, 0 | 1, 1, 1, 1, 1 <= array_id (generated from offsets, needed for subsequent unique_by_key)
---- UniqueByKey, will use a zip_iterator of array_id and post_sort_items
0, 2 | 3, 5, 7 <= unique_sort_indexes (sort_indexes of the unique items)
0, 1 | 2, 3, 4 <= dictionary_entry
---- Scatter: for i=0..total_unique_items: prepared_output[unique_sort_indexes[i]] = dictionary_entry[i];
0  1  2 | 3  4  5  6  7 <= indexes
0, -, 1 | 2, -, 3, -, 4 <= prepared_output (within the sorted order, what's the index of the )
---- Prefix-scan that fills in the blanks, sweeping the latest non-blank item to the right
0, 0, 1 | 2, 2, 3, 3, 4 <= filled_prepared_output
---- Now we have this:
0, 0, 1 | 2, 2, 3, 3, 4 <= filled_prepared_output
0, 2, 1 | 3, 7, 5, 6, 4 <= original_item_indexes (from the indexes after segmented sort)
---- Then Scatter back: for i=0..len: output[original_item_indexes[i]] = filled_prepared_output[i]
0  1  2 | 3  4  5  6  7 <= indexes
0, 1, 0 | 2, 4, 3, 3, 2 <=  output
```

As you can see, it requires a couple of cub algorithms. If this is an approach you'd be interested in, let me know and I can see if I can support in translating this to the respective cub algorithm invocations. However, I am not sure when I will get to that.

**Reply by wanghongyu2001** [reply]
Yes, this is exactly what I want. Thank you!

For convenience, let me describe the problem again. Please don't feel discouraged by my verbosity.

I have hundreds of thousands of segments, each with a size ranging from 0 to several hundred thousand. Each segment is independent and not related to others. They are only adjacent in storage. So they have high parallelism.

E.g.
```
0, 3, 8                 <= offset
0  1  2 | 3  4  5  6  7 <= indexes
0, 4, 0 | 1, 4, 2, 2, 1 <= input
0, 1, 0 | 0, 2, 1, 1, 0 <= output
```
